### DISCUSSION TITLE
(Lazarus / FreePascal) Reading an image and assigning pixels to 16x16 grid?

### DISCUSSION THREAD

**Discussion by FiftyTifty** [ImageMagick/ImageMagick#6343] [Help] [answered]
### ImageMagick version

ImageMagick-6.9.12-Q16-HDRI

### Operating system, version and so on

Windows 10 x64 LTSC

### Description

Been using of ImageMagick via command line, and now wanting to try using it in software, to automate a really tedious process.

Here's what I'm trying to achieve:
- Read a .png image
- Read image as an array of 16x16 pixels
- If any pixel is not pure black
- * Make the whole 16x16 tile pure white
- After iterating through all the pixels, save the new image

The documentation for ImageMagick is really bad on Lazarus, it's just a code example. No explanation of what things do, or what they are. It can be found here: https://wiki.freepascal.org/PascalMagick

I've done the process in Photoshop with an example image. Got an RGB image that is just white brush strokes on a black background. With a grid of 16x16 pixels per tile, use the mosaic filter (pixelize). Then threshold the result with a tolerance that is the smallest number possible above 0 (for 8bit RGB images, that number is 1 out of 255). The result is turning a soft mask, into a hard mask constrained to 16x16 groups of pixels.

Sounds complicated, but easier to show with screenshots, which I have uploaded to imgur:

Mask with photoshop grid (16x16px per tile) overlaid on top: https://i.imgur.com/MmC6dbv.png
Mask after using mosaic filter (pixels automatically constrained to grid): https://i.imgur.com/GQMdC1U.png
Mask after using threshold (clamp all non-0 values to 255): https://i.imgur.com/Oe5Afzz.png

How would I go about this? Probably picked a bad language as it's not used often these days, but it's easy, fast (as C#), and works seamlessly on all manner of operating systems without any effort on my or the user's part.

**Comment by fmw42** [comment]
Sorry, I do not know PascalMagick, but here is a way to do that in shell scripting in bash unix.

```
Read the image and crop into blocks of 16 and save as IM format MIFF
Loop over each block image and test if the mean (average) is larger than 0 (i.e. has non-black pixels)
If the test is non-black pixels, then write over the block image with white
Recombine all the block images to reform the full image. 
(Note: MIFF saves the virtual pixel offsets, so can flatten to recombine)
```

Input:

![crosses](https://github.com/ImageMagick/ImageMagick/assets/13458218/2908f5e2-4b30-448d-909b-aad60c4994a7)


```
cd
cd desktop/test1
magick crosses.gif -crop 16x16 crosses_%02d.miff
for img in *.miff; do
test=$(magick "$img" -format "%[fx:u.mean>0?1:0]" info:)
[ $test -eq 1 ] && magick "$img" -fill white -colorize 100 "$img"
done
magick *.miff -background none -flatten crosses_result.png

```

![crosses_result](https://github.com/ImageMagick/ImageMagick/assets/13458218/0e0788ee-103f-4125-bd49-2be68030c9e9)


See https://imagemagick.org/Usage/crop/#crop_tile

**Reply by FiftyTifty** [reply]
Oh that's really interesting. Thanks for making that. If there was a way to run .sh scripts on Windows without Git, that would be the solution.

PascalMagick is just Pascal bindings for the C stuff. Apparently done through the use of MagickWand functions?

**Reply by fmw42** [reply]
But you should be able to write code to emulate what I have supplied in pseudocode.

**Reply by fmw42** [reply]
Why must it be Pascal?

**Reply by FiftyTifty** [reply]
Pascal is the language I'm most comfortable with, it's multiplatform with native GUI for every OS (fpgui is amazing, my tools need to run on Windows, Linux, and MacOS), and has good performance.

The commands aren't unique to PascalMagick either, as it is just bindings to the C code for ImageMagick. That's what ImageMagick's MagickWand API is for. What are the commands used to achieve the same as your script?

**Comment by snibgo** [comment]
If the width and height of the input image are both multiples of 16, the task can be done in one simple IM command. `-scale` it down to exactly `w/16` by `h/16`. "`-fill White +opaque Black`" (so black stays black and anything else becomes white). Scale it back up to the original size.

If the width or height is not a multiple of 16, your description doesn't say what should happen.

Incidentally, your example images do not include the actual input. So I can't compare the result from my simple IM with your expected result.

From the page at https://wiki.freepascal.org/PascalMagick , the IM functions seem to be exactly those in https://github.com/ImageMagick/ImageMagick/blob/main/MagickWand/magick-image.h . See the `.c` file for comments on the functions.

**Reply by FiftyTifty** [reply]
Ooh that is really cool. Yeah the input maps will always be divisible by 16 (masks are typically 256, 512, 1024, 2048, etc) And my apologies for not sharing the example image, I forgot to share one. I've quickly made another mock up, 
![Untitled](https://github.com/ImageMagick/ImageMagick/assets/13511114/e0f6f705-b123-47e9-a218-11392e161363)

Yeah, PascalMagick uses the exact same ones as MagickWand. That's great, I'll check that out, and report back how I get on with this.

**Comment by snibgo** [comment]
> If there was a way to run .sh scripts on Windows without Git, that would be the solution.

Microsoft has something for that. Personally, I use bash from Cygwin.

**Comment by snibgo** [comment, marked as answer]
Where `inp16x16.png` is your input file, a Windows BAT command is:
```
magick ^
  inp16x16.png ^
  -set option:MYSIZE %%[fx:w]x%%[fx:h] ^
  -scale "%%[fx:w/16]x%%[fx:h/16]^!" ^
  -fill White +opaque Black ^
  -scale "%%[MYSIZE]^!" ^
  x.png
```
The result, `x.png`, is:

![x](https://github.com/ImageMagick/ImageMagick/assets/33812368/7a6cf572-57b0-42cb-a4bf-0e3bcc05b08b)

I can't see any advantage in making this into a compiled program. The performance would be the same.

Some operations require us to loop through multiple images in order to make one output. And some operations require "`if ... then ... else`" conditions. These operations are more conveniently done within a compiled program (compared to a shell script), and may give increased performance because a program may not need to save temporary files.

**Reply by FiftyTifty** [reply]
The main reason is to keep it as a self contained tool. So the end user doesn't need to have ImageMagick installed, and to allow for drag-drop to automatically process the image. But honestly, these .bat files you two have been sharing do the job. I really appreciate you making these, as I wouldn't have been able to figure this out myself.
